openvino/itt/include/openvino/itt.hpp

```diff
@@ -0,0 +1,108 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+/**
+ * @brief Defines API to trace using Intel ITT.
+ * @file itt.hpp
+ */
+
+#pragma once
+#include "itt_domains.hpp"
+#include <string>
+
+namespace openvino
+{
+    namespace itt
+    {
+        typedef struct domain_ {} *domain_t;
+        typedef struct handle_ {} *handle_t;
+
+        namespace internal
+        {
+            domain_t domain(char const* name);
+            handle_t handle(char const* name);
+            void taskBegin(domain_t d, handle_t t);
+            void taskEnd(domain_t d);
+            void threadName(const char* name);
+        }
+
+        inline void threadName(const char* name)
+        {
+            internal::threadName(name);
+        }
+
+        inline void threadName(const std::string &name)
+        {
+            internal::threadName(name.c_str());
+        }
+
+        inline handle_t handle(char const *name)
+        {
+            return internal::handle(name);
+        }
+
+        inline handle_t handle(const std::string &name)
+        {
+            return internal::handle(name.c_str());
+        }
+
+        template <typename Tag>
+        domain_t domain()
+        {
+            static auto d = internal::domain(Tag::name());
+            return d;
+        }
+
+        template <typename Tag>
+        handle_t handle()
+        {
+            static auto h = internal::handle(Tag::name());
+            return h;
+        }
+
+        template <typename DomainTag>
+        struct ScopedTask
+        {
+            ScopedTask(handle_t taskHandle) noexcept
+            {
+                internal::taskBegin(domain<DomainTag>(), taskHandle);
+            }
+            ~ScopedTask() noexcept { internal::taskEnd(domain<DomainTag>()); }
+            ScopedTask(const ScopedTask&) = delete;
+            ScopedTask& operator=(const ScopedTask&) = delete;
+        };
+
+#define OV_ITT_DOMAIN(domainName)                                                                   \
+        struct domainName                                                                           \
+        {                                                                                           \
+            static const char * name() noexcept                                                     \
+            {                                                                                       \
+                return #domainName;                                                                 \
+            }                                                                                       \
+        };
+
+#define OV_ITT_SCOPED_TASK(domain, taskName)                                                        \
```

Comment:
Some simplification proposal:
```c++
       template <int>
        inline handle_t handle(const char* name)
        {
            static auto h = internal::handle(name);
            return h;
        }

        template <int>
        inline handle_t handle(handle_t h)
        {
            return h;
        }

        template<domain_t(*domain)()>
        struct ScopedTask
        {
            ScopedTask(handle_t taskHandle) noexcept
            {
                internal::taskBegin(domain(), taskHandle);
            }
            ~ScopedTask() noexcept { internal::taskEnd(domain()); }
        };

#define OV_ITT_DOMAIN(domainName)                                                                   \
    static domain_t domainName() noexcept                                                     \
    {                                                                                       \
        static auto d = internal::domain(#domainName);
        return d;                                                                \
    }                                                                                       \

#define OV_ITT_SCOPED_TASK(domain, taskOrTaskName)                                                        \
    openvino::itt::ScopedTask<domain> ittScopedTask ## __LINE__  {openvino::itt::handle<__LINE__>(taskOrTaskName)};
```

So we can use one `OV_ITT_SCOPED_TASK` for all cases.

Reply:
Unfortunately we can't use only line number as template parameter for handle() because template argument for handle() should be unique. Otherwise we can use the same ITT counter for different tasks in different modules.

Reply:
What about:
```c++
       template <typename>
        inline handle_t handle(const char* name)
        {
            static auto h = internal::handle(name);
            return h;
        }

        template <typename>
        inline handle_t handle(handle_t h)
        {
            return h;
        }

        template<domain_t(*domain)()>
        struct ScopedTask
        {
            ScopedTask(handle_t taskHandle) noexcept
            {
                internal::taskBegin(domain(), taskHandle);
            }
            ~ScopedTask() noexcept { internal::taskEnd(domain()); }
        };

#define OV_ITT_DOMAIN(domainName)                                                                   \
    static domain_t domainName() noexcept                                                     \
    {                                                                                       \
        static auto d = internal::domain(#domainName);
        return d;                                                                \
    }                                                                                       \

#define OV_ITT_SCOPED_TASK(domain, taskOrTaskName)                                                        \
    openvino::itt::ScopedTask<domain> ittScopedTask ## __LINE__  {openvino::itt::handle<struct TaskTag>(taskOrTaskName)};
```

Reply:
With line certainly:
```c++
#define OV_ITT_SCOPED_TASK(domain, taskOrTaskName)                                                        \
    openvino::itt::ScopedTask<domain> ittScopedTask ## __LINE__  {openvino::itt::handle<struct TaskTag##__LINE__>(taskOrTaskName)};
```

Reply:
Yes it's better but I'm not sure about "struct TaskTag". (:
We have non zero chance to clash with the real definition in other module.
Also I'm not sure, that "struct TaskTag" generates different istantiations for handle() in different calls. I'll check it (:

Reply:
> struct TaskTag##__LINE__
I think, If we call OV_ITT_SCOPED_TASK in different files at the same line, we get the same instantiation for template function. We have the same probleam as with __LINE__ as template argument. If we somehow add file name it should work. (:

Reply:
No `struct TaskTag` will be local structure: https://godbolt.org/z/nG7er8

Reply:
`__LINE__` is used for different tasks in the same function

Reply:
Yes, if lines are different we don't have any issues. We can face with problem if we call OV_ITT_SCOPED_TASK at the same line in different modules.

For instance:
In `foo.cpp:10` we get this one:` itt::handle<struct TaskTag10>`.
In `boo.cpp:10` we get the same instantiation: `itt::handle<struct TaskTag10>`.

Reply:
for foo.cpp:
```c++
 void foo() {
    itt::handle<struct TaskTag10>.
}
```

for bar.cpp:
```c++
 void bar() {
    itt::handle<struct TaskTag10>.
}
```

Compiler will generate foo()::TaskTag10 and bar()::TaskTag10 local struts names.

Reply:
Ok, I just checked. Seems It works (:
But I think the explicit type declaration is better because we can avoid types clashes in further. Somebody can define struct TaskTag10 somewhere. Thanks a lot for improvements. I'll change macros according your proposal.